### PR TITLE
[v2.0.0] Wallets v2 & 2FA & Floating Fees

### PR DESCRIPTION
**Pull request by rubensayshi**
v2.0.0
------
 - New [Default] Wallet Version 2
 - `BackupGenerator` now supports `$extra` to be printed on document for extra notes
 - 2FA support
 - Floating Fees

### Upgrade / BC breaks
 - `createNewWallet` now returns `list($wallet, $backupInfo)` to be able to support both v1 and v2
 - `new BackupGenerator()` now takes `$identifier, $backupInfo, $extra`
 - No longer support `BackupGenerator::generateImage`
 - `WalletSweeper` has changed to remove some abstraction, also added insight-api implementation
 - Defaults to using `Wallet::FEE_STRATEGY_OPTIMAL` instead of `Wallet::FEE_STRATEGY_BASE_FEE`!

### New [Default] Wallet Version 2
Instead of using `BIP39`, wallet seeds will now be stored encrypted - to allow for password changes

Wallet Creation:  
```
primarySeed = random()
secret = random()
primaryMnemonic = BIP39.entropyToMnemonic(AES.encrypt(primarySeed, secret))
secretMnemonic = BIP39.entropyToMnemonic(AES.encrypt(secret, password))
```

Wallet Init:  
```
secret = BIP39.entropyToMnemonic(AES.decrypt(secretMnemonic, password))
primarySeed = BIP39.entropyToMnemonic(AES.decrypt(primaryMnemonic, secret))
```

See `docs/KEYS.md` for more info
   
Old Wallets that are v1 will remain so and will continue working.

Wallets created through the Blocktrail website will have an option to choose which version to create (default v2), 
as grace period the first few days after the SDK release this will default to v1 still.

### Floating Fees
To deal with stresstests etc. there's now a `$feePerKB` method to get the optimal fee and the `$wallet->pay` has a `$feeStrategy` argument.  
When `$feeStrategy` is `Wallet::FEE_STRATEGY_OPTIMAL` (default) it will use the (by the API calculated) optimal fee per KB.  
When `$feeStrategy` is `Wallet::FEE_STRATEGY_BASE_FEE` it will use the BASE_FEE of 0.0001 BTC per KB and use the old way of rounding the transaction size to 1 KB.

For using the `TransactionBuilder` use the `$txBuilder->setFeeStrategy()` if you don't want the default `Wallet::FEE_STRATEGY_OPTIMAL`.

Optimal fee is calculated by taking all transactions from the last 30 blocks and calculating what the lowest possible fee is 
that still gives more than 80% chance to end up in the next block.

### CODE REVIEW

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,60 @@
...
+
+### New [Default] Wallet Version 2
+Instead of using `BIP39`, wallet seeds will now be stored encrypted - to allow for password changes
+
+Wallet Creation:  
+```
+primarySeed = random()
+secret = random()
+primaryMnemonic = BIP39.entropyToMnemonic(AES.encrypt(primarySeed, secret))
+secretMnemonic = BIP39.entropyToMnemonic(AES.encrypt(secret, password))
```
Comment (OACDesigns) [inline docs/CHANGELOG.md]: are we calling it "secret mnemonic" now instead of backup mnemonic?

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,60 @@
 BlockTrail PHP SDK Changelog
 ============================
 
+v2.0.0
+------
+ - New [Default] Wallet Version 2
```
Comment (OACDesigns) [inline docs/CHANGELOG.md]: should probably mention that new wallets support password changing (the main reason we changed it)

File: src/BackupGenerator.php [hunk 3]
```diff
@@ -119,251 +112,89 @@ public function generateHTML() {
...
-            ['font-size' => 18, 'line-height' => 26, 'text' => 'Backup Info'],
-            ['font-size' => 14, 'line-height' => 24, 'text' => 'Primary Mnemonic: '],
-        );
-        //chunk the mnemonics into groups of x words
-        preg_match_all('/(\w*\s){1,8}/', $this->primaryMnemonic, $results);
-        foreach ($results[0] as $wordGroup) {
-            $backupContent[] = array('font-size' => 12, 'line-height' => 20, 'text' => $wordGroup);
+        if (isset($this->backupInfo['backup_mnemonic'])) {
+            $backupInfo[] = [
+                'title' => 'Primary Mnemonic',
```
Comment (OACDesigns) [inline src/BackupGenerator.php]: should be "Backup Mnemonic"?

**Comment by rubensayshi** [comment]
closing, squashed in https://github.com/blocktrail/blocktrail-sdk-php/pull/30 and better to release it in 1 go with all the bitcoin-lib upgrade changes.

**Pull request closed without merging**